flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/CastFunctionITCase.java

```diff
@@ -18,176 +18,905 @@
-                                "RowToFirstField(StructuredTypeConstructor((f0, f1)))",
-                                12,
-                                DataTypes.INT()),
-                TestSpec.forFunction(
-                                BuiltInFunctionDefinitions.CAST,
-                                "explicit between structured type and row")
-                        .onFieldsWithData(12, "Ingo")
-                        .withFunction(StructuredTypeConstructor.class)
-                        .testTableApiResult(
-                                call("StructuredTypeConstructor", row($("f0"), $("f1")))
-                                        .cast(
-                                                DataTypes.ROW(
-                                                        DataTypes.BIGINT(), DataTypes.STRING())),
-                                Row.of(12L, "Ingo"),
-                                DataTypes.ROW(DataTypes.BIGINT(), DataTypes.STRING())));
+                CastTestSpecBuilder.test(CHAR(3), "To CHAR(3)")
+                        .testCase(CHAR(3), "foo", "foo")
+                        .testCase(CHAR(4), "foo", "foo ")
+                        .testCase(CHAR(4), "foo ", "foo ")
+                        .testCase(VARCHAR(3), "foo", "foo")
+                        .testCase(VARCHAR(5), "foo", "foo")
+                        .testCase(VARCHAR(5), "foo ", "foo ")
+                        // https://issues.apache.org/jira/browse/FLINK-24413 - Trim to precision
+                        // in this case down to 3 chars
+                        .testCase(STRING(), "abcdef", "abcdef") // "abc"
+                        .testCase(DATE(), LocalDate.parse("2021-30-09"), "2021-30-09") // "202"
+                        .build(),
+                CastTestSpecBuilder.test(VARCHAR(3), "To VARCHAR(3)")
+                        .testCase(CHAR(3), "foo", "foo")
+                        .testCase(CHAR(4), "foo", "foo ")
+                        .testCase(CHAR(4), "foo ", "foo ")
+                        .testCase(VARCHAR(3), "foo", "foo")
+                        .testCase(VARCHAR(5), "foo", "foo")
+                        .testCase(VARCHAR(5), "foo ", "foo ")
+                        // https://issues.apache.org/jira/browse/FLINK-24413 - Trim to precision
+                        // in this case down to 3 chars
+                        .testCase(STRING(), "abcdef", "abcdef")
+                        .build(),
+                CastTestSpecBuilder.test(STRING(), "To STRING")
+                        .testCase(CHAR(3), "foo", "foo")
+                        .testCase(CHAR(5), "foo", "foo  ")
+                        .testCase(VARCHAR(5), "Flink", "Flink")
+                        .testCase(VARCHAR(10), "Flink", "Flink")
+                        .testCase(STRING(), "Apache Flink", "Apache Flink")
+                        .testCase(STRING(), null, null)
+                        .testCase(BOOLEAN(), true, "true")
```

Comment:
this should be uppercase: `TRUE`

Reply:
Currently, it's lowercase, I guess it will be fixed with you current work?

Reply:
I don't think my fix will influence this code path. Does `CAST("true" AS BOOLEAN)` works? Or it requires `TRUE` uppercase?

Reply:
the inverse works as case insensitive, if you look at the tests listed:
```
.fromCase(CHAR(3), "foo", null)
.fromCase(CHAR(4), "true", true)
.fromCase(VARCHAR(5), "FalsE", false)
.fromCase(STRING(), "Apache Flink", null)
.fromCase(STRING(), "TRUE", true)
```